services/cannon/src/Cannon/WS.hs

```diff
@@ -234,6 +237,23 @@ sendMsgIO :: L.ByteString -> Websocket -> IO ()
 sendMsgIO m c = do
   recoverAll retry3x $ const $ sendBinaryData (connection c) m
 
+sendMsgConduit :: Key -> Websocket -> ConduitT ByteString Void (ResourceT WS) ()
+sendMsgConduit k c = do
+  m <- await
+  case m of
+    Just m' -> do
+      lift $ traceLog m'
+      liftIO $ sendToWebsocket m'
+    Nothing -> pure ()
+  where
+    traceLog :: ByteString -> (ResourceT WS) ()
+    traceLog m = lift $ trace $ client kb . msg (val "sendMsg: \"" +++ Data.ByteString.take 128 m +++ val "...\"")
+
+    kb = key2bytes k
+
+    sendToWebsocket :: ByteString -> IO ()
```

Comment:
You could make `sendMsgIO` polymorphic in the input type (i.e. take an `a` constrained by `WebSocketsData a`), and use that instead of redefining it here.

Reply:
:+1: